src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -201,57 +111,86 @@ void SingleThreadSampler::sample_single_variable(long vid, bool is_inc) {
-        if ((*this->p_rand_obj_buf) *
-                (1.0 + exp(potential_neg - potential_pos)) <
-            1.0) {
-          p_fg->template update<false>(variable, 1.0);
-        } else {
-          p_fg->template update<false>(variable, 0.0);
-        }
+      *this->p_rand_obj_buf = erand48(this->p_rand_seed);
+      // sample the variable
+      // flip a coin with probability
+      // (exp(potential_pos) + exp(potential_neg)) / exp(potential_neg)
+      // = exp(potential_pos - potential_neg) + 1
+      if ((*this->p_rand_obj_buf) * (1.0 + exp(potential_neg - potential_pos)) <
+          1.0) {
+        proposal = 1;
+      } else {
+        proposal = 0;
       }
+      break;
 
-    } else if (variable.domain_type == DTYPE_MULTINOMIAL) {
+    case DTYPE_MULTINOMIAL:
+      // varlen_potential_buffer contains potential for each proposals
       while (variable.upper_bound >= varlen_potential_buffer.size()) {
         varlen_potential_buffer.push_back(0.0);
       }
 
-      if (variable.is_evid == false || sample_evidence) {
-        sum = -100000.0;
-        acc = 0.0;
-        multi_proposal = -1;
-        for (int propose = variable.lower_bound;
-             propose <= variable.upper_bound; propose++) {
+      sum = -100000.0;
+      acc = 0.0;
+      proposal = -1;
+
+      // calculate potential for each proposal
+      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
+           propose++) {
+        if (is_free_sample) {
+          varlen_potential_buffer[propose] =
+              p_fg->template potential<true>(variable, propose);
+        } else {
           varlen_potential_buffer[propose] =
               p_fg->template potential<false>(variable, propose);
-          sum = logadd(sum, varlen_potential_buffer[propose]);
         }
+        sum = logadd(sum, varlen_potential_buffer[propose]);
+      }
 
-        *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-        for (int propose = variable.lower_bound;
-             propose <= variable.upper_bound; propose++) {
-          acc += exp(varlen_potential_buffer[propose] - sum);
-          if (*this->p_rand_obj_buf <= acc) {
-            multi_proposal = propose;
-            break;
-          }
+      // flip a coin
+      *this->p_rand_obj_buf = erand48(this->p_rand_seed);
+      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
+           propose++) {
+        acc += exp(varlen_potential_buffer[propose] - sum);
+        if (*this->p_rand_obj_buf <= acc) {
+          proposal = propose;
+          break;
         }
-        assert(multi_proposal != -1);
-        p_fg->template update<false>(variable, multi_proposal);
       }
 
-    } else {
+      assert(proposal != -1);
+      break;
+
+    default:
       std::cerr
           << "[ERROR] Only Boolean and Multinomial variables are supported now!"
           << std::endl;
-      assert(false);
-      return;
-    }
+      abort();
```

Comment:
I think we can drop the `cerr` messages and just put a comment here.. since you said it's already done by the input loading code.